plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
+        $visitFieldsToSelect = array_merge(array('idvisit', 'location_ip'), array_keys(self::$logVisitFieldsToUpdate));
+
+        $lastId = 0;
+        do {
+            $logs = $this->dao->getVisitsWithDatesLimit($from, $to, $visitFieldsToSelect, $lastId, $segmentLimit);
+            if (!empty($logs)) {
+                $lastId = $logs[count($logs) - 1]['idvisit'];
+
+                $this->reattributeVisitLogs($output, $logs);
+            }
+        } while (count($logs) == $segmentLimit);
+    }
+
+    protected function reattributeVisitLogs(OutputInterface $output, $logRows)
+    {
+        foreach ($logRows as $row) {
+            if (!$this->isRowComplete($output, $row)) {
+                continue;
+            }
+
+            $location = $this->getVisitLocation($row);
+            $valuesToUpdate = $this->getValuesToUpdate($row, $location);
+
+            $this->onVisitProcessed($output);
+
+            $idVisit = $row['idvisit'];
+
+            if (empty($valuesToUpdate)) {
+                $this->writeIfVerbose($output, 'Nothing to update for idvisit = ' . $idVisit . '. Existing location info is same as geolocated.');
+            } else {
+                $this->writeIfVerbose($output, 'Updating visit with idvisit = ' . $idVisit . '.');
+
+                $this->dao->updateVisits($valuesToUpdate, $idVisit);
+                $this->dao->updateConversions($valuesToUpdate, $idVisit);
+            }
+        }
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return int
+     */
+    protected function getPercentStep(InputInterface $input)
+    {
+        $percentStep = $input->getArgument(self::PERCENT_STEP_ARGUMENT);
+
+        if (!is_numeric($percentStep)) {
+            return self::PERCENT_STEP_ARGUMENT_DEFAULT;
+        }
+
+        // Percent step should be between maximum percent value and minimum percent value (1-100)
+        if ($percentStep > 99 || $percentStep < 1) {
+            return 100;
+        }
+
+        return $percentStep;
+    }
+
+    /**
+     * Returns location log values that are different than the values currently in a log row.
+     *
+     * @param array $row The visit row.
+     * @param array $location The location information.
+     * @return array The location properties to update.
+     */
+    protected function getValuesToUpdate(array $row, $location)
```

Comment:
> There will be a new class that may never be re-used or meant to be re-used, that if it is ever re-used can be easily used in an improper context or can have its logic re-implemented because it is not flexible enough, that will have tests that need to be maintained but only test what is a seldom used and simple administrative function.

We do not only create a class to have it reusable. This was not my point here. As said I prefer to have a class for better readable code, for clean code and for testable code. 

> There will be a new class that may never be re-used or meant to be re-used, that if it is ever re-used can be easily used in an improper context or can have its logic re-implemented because it is not flexible enough, that will have tests that need to be maintained but only test what is a seldom used and simple administrative function.

Whether a class is ever re-used is not really important see comment above. If someone uses a class in an improper context or wants to make a class more flexible and ends up in a big mess then this should be detected in a code review here as it would be most likely a hack. This was often the case in the past but I think we're avoiding such things now (hopefully). If the class is not flexible enough we hopefully refactor the code in a way that it is good design but not over-engineered. Of course it requires a lot of discipline and I can understand your argument. 

> Since PHP + Symfony forces the metadata to be defined procedurally, and encourages the actual logic to be hidden, we run into this problem.

I don't think Symfony encourages the actual logic to be hidden. I reckon the idea there is to have the logic somewhere else outside the command and then it is not a problem. But it probably always depends on the type of command / application etc. We can at some point rethink about different solutions in general that also include controllers. So that we in the end can decide what should be available via a controller (web), via a command, or both, etc. That's a different topic though :)

> I would not consider this a good way to resolve a difference of opinion. I would prefer to reach an understanding, rather than one person acting independently or capitulating unconditionally.

Matt explained it quite well. It's not super important to me and it was only a suggestion. It's nothing were I would insist on something and to me the benefit is not worth a huge discussion.

> I think the current goal should be to have something usable in 2.12 for the particular client who has been waiting for a long time to have this functionality in cloud. 

I think a client should have something where we can be pretty sure that it actually works. If we don't know whether it actually works, we should not ship it. If a method does not have unit/integration tests for single methods then it will most likely have bugs. Even if one things it must/should work. I notice this at least in my code all the time and whenever I write a unit / integration test for a method I do find bugs. 100% guaranteed. I've been working on the Sort filter for the last 2 days and even there adding tests for each method allowed me to find many bugs where I never thought there would be a bug. I usually even make the experience that by having tests for each method I end up writing less tests and I'm faster this way since I don't have to write system tests that cover all possible ways. Also we most likely won't add unit/integration tests once a feature is "done" and shipped ;)